### PR TITLE
Iotree limit

### PR DESCRIPTION
**Pull request by waychal**
It adds the in_limit and out_limit parameters for io_tree endpoint of rest api. If limit is not specified, it will return all the incoming and outgoings of the requested node.

URL: http://127.0.0.1:5000/api/v2/calculations/78e496d2/io/tree?in_limit=1&out_limit=1

This feature requires the pagination but it will be handled in separate pull request.

### CODE REVIEW

**Comment by ltalirz** [comment]
@giovannipizzi @sphuber  Snehal sees some [strange transport.plugins test failures](https://travis-ci.org/waychal/aiida_core/jobs/489976559#L1389)  although she seemingly hasn't touched any code related to this.

Do you guys happen to have an idea what this could be?

**Comment by sphuber** [comment]
We have the same on `provenance_redesign` and I have no idea what is going on but it seems to be a change that happened on Travis. I tried to diff the logs with a build that worked but could see nothing that caught my eye. I have already sent an email this morning to Travis support but have no response yet.

**Comment by giovannipizzi** [comment]
yes, @sphuber already pinged the Travis guys, we believe it's due to a change on their side

**Comment by ltalirz** [comment]
Thanks for the quick feedback!

**Comment by giovannipizzi** [comment]
I think that in general it is better to have a "hardcoded" max in the server; this makes the client a bit more complex, having to deal with pagination, but avoids that the client (by mistake/naively, not specifying any parameter, or intentionally specifying a very large number) creates requests that are very expensive and can block the server. What do you think?

**Comment by ltalirz** [comment]
We agree but we (@waychal and me) decided that since this PR already solves the main problem, pagination (and a default maximum) would be added in a separate PR later, allowing @waychal to work on the phonon visualizer instead.

File: aiida/restapi/resources.py [hunk 1]
```diff
@@ -259,9 +261,10 @@ def get(self, id=None, page=None):
             results = self.trans.get_statistics(usr)
 
         # TODO Might need to be improved
```
Comment (ltalirz) [inline aiida/restapi/resources.py]: The first todo is too general to be useful. Can it be removed?

File: aiida/restapi/resources.py [hunk 2]
```diff
@@ -259,9 +261,10 @@ def get(self, id=None, page=None):
...
+        # TODO add pagination for this endpoint
```
Comment (ltalirz) [inline aiida/restapi/resources.py]:
```suggestion
        # TODO add pagination for this endpoint (then add defaults for tree_in_limit, tree_out_limit)
```

File: aiida/restapi/translator/node.py [hunk 3]
```diff
@@ -668,5 +673,16 @@ def get_node_shape(ntype):
...
+        no_of_inputs = qb.count()
+
+        qb = QueryBuilder()
+        qb.append(Node, tag="main", project=['id'], filters=self._id_filter)
+        qb.append(Node, tag="out", project=['id'], output_of='main')
+        no_of_outputs = qb.count()
+
 
-        return {"nodes": nodes, "edges": edges}
+        return {"nodes": nodes, "edges": edges, "total_no_of_incomings": no_of_inputs, "total_no_of_outgoings": no_of_outputs}
```
Comment (ltalirz) [inline aiida/restapi/translator/node.py]: Is the `total_` prefix needed?
What about `no_of_incomings`?

Comment (giovannipizzi) [reply aiida/restapi/translator/node.py]: I think it's more common to use `num_` instead of `no_`

**Comment by ltalirz** [comment]
@waychal 
One question: If the JS graph explorer now simply starts using the `tree_in_limit` URL parameter, will this mean that the graph explorer no longer works with aiida versions < 0.12.3?
The usual way to avoid this would be to check the version of the API, are you currently doing this?
E.g. this is the output of the `/server` endpoint:
```
{
  "data": [
    "REST API version: v2", 
    "REST API Prefix: /api/v2", 
    "AiiDA==0.12.2"
  ], 
  "method": "GET", 
  "path": "/api/v2/server", 
  "query_string": "", 
  "resource_type": "Info", 
  "url": "http://127.0.0.1:5000/api/v2/server", 
  "url_root": "http://127.0.0.1:5000/"
}
```

Mentioning also @giovannipizzi for comment

**Comment by giovannipizzi** [comment]
Good point, also, are we increasing the version number (and is this needed)? Or we just rely on the AiiDA version? Probably we need an API sub-version (that indeed could be represented by the AiiDA version)

**Comment by waychal** [comment]
@ltalirz I guess in old version of aiida, `tree_in_limit` URL parameter should simply be ignored and it will return the complete list of incoming and outgoing nodes but I need to check it in code. In Materials cloud both cases can be handled. So the rest api of all versions of aiida will work.

@giovannipizzi This is minor change in api so I would not suggest to increase the version number but its good idea to start to use the api sub-versions now.

**Comment by ltalirz** [comment]
@waychal I see - I just noticed that some endpoints fail when providing URL parameters they don't know - but you are right, the /tree endpoint seems to ignore them.

So, for the moment I'm ok without introducing this. We should probably keep it in mind for the future, though, as at some point we will encounter backwards-incompatible additions.

**Comment by ltalirz** [comment]
I also thought about this problem and came to the same conclusion.
There are not that many PRs in 0.12.3, so going through them and assigning the "aiida-core 0.x" label is quite doable.
The remaining ones then need to be cherry-picked.

**Pull request merged**